services/construction.go

```diff
@@ -622,82 +593,110 @@ func (s *constructionAPIService) ConstructionPayloads(
 	}
 	nonce := uint64(nonceF64)
 
-	if len(request.Operations) < 2 {
-		loggerCons.Error("ConstructionPayloads: missing fee operation")
-		return nil, ErrMalformedValue
-	}
-	feeOp := request.Operations[0]
-	if feeOp.Type != OpTransfer {
-		loggerCons.Error("ConstructionPayloads: fee operation wrong type",
-			"type", feeOp.Type,
-			"expected_type", OpTransfer,
-		)
-		return nil, ErrMalformedValue
-	}
-	if feeOp.Account.SubAccount != nil {
-		loggerCons.Error("ConstructionPayloads: fee operation wrong subaccount",
-			"sub_account", feeOp.Account.SubAccount,
-			"expected_sub_account", nil,
-		)
-		return nil, ErrMalformedValue
-	}
-	signWithAddr := feeOp.Account.Address
-	feeAmount, err := readOasisCurrencyNeg(feeOp.Amount)
-	if err != nil {
-		loggerCons.Error("ConstructionPayloads: readOasisCurrency feeOp.Amount",
-			"amount", feeOp.Amount,
-			"err", err,
-		)
-		return nil, ErrMalformedValue
-	}
+	remainingOps := request.Operations
+	var signWithAddr string
 	feeGas := DefaultGas
-	if feeGasRaw, ok := feeOp.Metadata[FeeGasKey]; ok {
-		feeGasF64, ok := feeGasRaw.(float64)
-		if !ok {
-			loggerCons.Error("ConstructionPayloads: malformed fee gas metadata")
+	feeAmount := quantity.NewQuantity()
+	var err error
+
+	if len(remainingOps) >= 2 &&
```

Comment:
Add comment explaining that we first match and pop the fee-related operations which are assumed to be the first ones.

We could also introduce a special OpFee operation so that we wouldn't rely on order? But then we would also need to convert that when parsing events which seems annoying unless we introduce a special staking event for fees or add a "kind" field to staking transfer events.

Reply:
thanks for the suggestion about adding the comment.

for the OpFee idea: I'm currently going about this under the philosophy that these intents are expressed in a network-specific "language" of sequences of operations where order is significant. And for some reason, the rest of the codebase seems not to prioritize being liberal in what it accepts